### PR TITLE
Adds recursive search and default author

### PR DESCRIPTION
**Pull request by amenadiel**
Adds a few enhancements to use the extractor
(assuming there is Python 3.x + bash + Makefile)

-  Point to a folder containing several repos, and pass `depth` parameter to search recursively 

```bash
./run.sh  --depth=2 ~/projects 
```
recursive_search gif

(when using the `depth` option the value of `output` is ignored and replaced with the folder name of each repo. This could be enhanced soon to instead create a folder named after the `output` parameter)

________________
- You can pass default email to preselect matching authors

```bash
./run.sh  --email=amenadiel@mail.com ~/projects/repo
```

preselect_emails2

________________
- Can skip auto upload prompt (if extracting several repos, you know...)

```bash
./run.sh  --upload=skip ~/projects/repo
```
skip_upload

________________________
 - Exception handling and exit early on edge cases
   - no branches
   - no commits
   - no authors
   - keyboard interrupt

_______________

The following changes were reverted and will be reconsidered, perhaps as an optional approach.

~~Everything runs on a virtualenv~~
~~make install installs python-venv and creates virtualenv (given you have python3) to avoid issues with versions and permissions~~
~~./run.sh activates the virtualenv when invoking src/main.py~~

### CODE REVIEW

**Comment by peti2001** [comment]
Hi @amenadiel,

Very nice pull request. I would recommend to split it into smaller PRs.
1.
 - Point to a folder containing several repos, and pass depth parameter to search recursively
 - You can pass default email to preselect matching authors
 - Can skip auto-upload prompt (if extracting several repos, you know...)
2.
- Using virtual envs

1: can be merged. These are really cool improvements. 
2: this is breaking change for me because of this in the `run.sh`:
```
bash -c "source ./.pyenv/bin/activate; python src/main.py $other_args \"$repostring\" " ;\
``` 
I am already using virtual env from `anaconda` and I might not want to install a new venv.
```
sudo apt install -y -q python3-venv
```
This change works only with `apt` package manager so CentOS, OSX, etc are excluded. Which I don't really like. 
I think it worth discussing, but in another PR and on the meanwhile let the users enjoy the other improvements you have created.

**Comment by amenadiel** [comment]
ok I will arrange stuff to avoid venv

**Comment by amenadiel** [comment]
@peti2001 

I removed the Makefile tasks and everything regarding my particular choice of virtualenv.

**Comment by amenadiel** [comment]
Perhaps I could include a `tasks.py` script and put in the README.md the optional approach of using [invoke](http://www.pyinvoke.org/) to run everything, like:

- install python3 with your tool of choice (apt, yum, brew, etc)
- install pip with your tool of choice
  - e.g. :  `wget https://bootstrap.pypa.io/get-pip.py && python3 get-pip.py --user`
- run `pip3 install invoke --user`
- run `invoke install` to create a virtual environment and install requirements in it
- run (something like) `invoke collect --depth=2 --email=me@me.com ~/repos`

(I'm just thinking at invoke because it expects double dashed parameters, unlike Makefile and fabric3)

**Comment by amenadiel** [comment]
@peti2001 ping

**Comment by peti2001** [comment]
Sorry for the late reply. Great work @amenadiel, merged to master!

TL;DR: Don't bother with virtual env, and pip install. We decided earlier it is not our responsibility to install Python, pip and the environments, because it is a very difficult topic.

There are many ways and solutions on how to set up your Python environment and we don't want to introduce an N+1 version. 
Instead of this, we should guide the user to finish these resources. Or if the user doesn't want to struggle setting up the environment help them to with other tools like: prepared Docker images and GitLab, GitHub integrations.
